### ISSUE TITLE
ENH: support Period.__contains__

### ISSUE THREAD

**Issue by jreback** [pandas-dev/pandas#15663]
supporting ``in`` for a ``Timestamp`` in a ``Period`` is straightforward and useful. Note that we current consider both endpoints inclusive.

```
In [28]: p = Period('2016',freq='M')

In [29]: p
Out[29]: Period('2016-01', 'M')

In [30]: p.start_time
Out[30]: Timestamp('2016-01-01 00:00:00')

In [31]: p.end_time
Out[31]: Timestamp('2016-01-31 23:59:59.999999999')

In [32]: p.start_time in p
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-32-b4a9c4196274> in <module>()
----> 1 p.start_time in p

TypeError: argument of type 'Period' is not iterable
```

**Comment by jreback** [comment]
cc @MaximilianR @shoyer

**Comment by shoyer** [comment]
I'm -0 on this one. `Period` is a scalar (not an iterable), which makes supporting Python's `in` a little surprising.

I would rather use a dedicated method to clear up any ambiguity, e.g., `p.contains(start_time)`.

**Comment by MaximilianR** [comment]
I'm -0.1. 

It depends whether `in` really means "in a collection" or we can override it to mean "in this shape". 

`.contains` is closer and is a decent start; easier to later override `__contains__` if we're not sure now

**Comment by jreback** [comment]
you guys are harsh! 

ok, as seen in #15309 I will have to refactor this a tiny bit.

as we use ``key in index`` to see if we have a valid key, so need to do something about this.

**Comment by holymonson** [comment]
Hi there, may I draw a little bit of your attention on this one? Either `Period.__contains__` or `Period.contains` or even `Timestamp.isin` would be fine, let's make one of them work.

**Comment by jreback** [comment]
@holymonson you are welcome to submit a PR; that's how things get fixed

**Comment by holymonson** [comment]
@jreback I'd like to submit a PR, the implement is quite simple, but first someone should decide which method name to be used.

**Comment by zerothi** [comment]
This would be ideal to have. While I see `__contains__` doesn't fit the iterable, it surely makes sense reading the code `t1 in period` makes perfect linguistic sense. However, it goes against the `__contains__` philosophy.

One could argue that iterating a period with any sub-frequency is allowed.

But it really needs to be in, one way or another! :)

Bump on this :)

**Comment by simonjayhawkins** [comment]
see https://github.com/pandas-dev/pandas/issues/15663#issuecomment-737888657

@zerothi  you are welcome to submit a PR; that's how things get fixed

**Comment by zerothi** [comment]
> see [#15663 (comment)](https://github.com/pandas-dev/pandas/issues/15663#issuecomment-737888657)
> 
> @zerothi you are welcome to submit a PR; that's how things get fixed

Yes, but please recognize https://github.com/pandas-dev/pandas/issues/15663#issuecomment-737982702 which asks for the convention? So which would be accepted?

**Comment by simonjayhawkins** [comment]
my read of the issue is that `.__contains__` would be rejected during review. so maybe `.contains`?

The implementation of those 2 would be moreless the same. so could easily change during review.

The issue title is about a method on Period. The other option of `Timestamp.isin` would probably have a significantly different implementation.

I think a PR for a `Period.contains` would be a good start.

'Period' object has no attribute 'contains' so no issue there.

**Comment by zerothi** [comment]
Why not decide prior to doing stuff? Discussing and settling that here seems appropriate.

**Comment by simonjayhawkins** [comment]
> Why not decide prior to doing stuff? Discussing and settling that here seems appropriate.

That would be ideal, but PRs tend to get more review/discussion/scrutiny than issues, so opening a PR has more chance of getting things moving than bumping an issue.
